File: experiment2/datalayer/src/main/scala/se/alejo720/experiment1partb/datalayer/DataLayer.scala
package se.alejo720.experiment1partb.datalayer
import java.io.File
import java.time.{Duration, Instant}

import com.github.tototoshi.csv._

import scala.collection.mutable

class Movie(ctitle: String, cgenres: List[String], cavgScore: Float) {
  var title: String = ctitle
  var genres: List[String] = cgenres
  var avgScore: Float = cavgScore
}

class DataLayer {
  val starts = Instant.now()
  val path = "C:\\Users\\AlexanderJ\\Google Drive\\Universitet\\Exjobb\\intellijidea-projects\\experiment1partB\\datalayer\\src\\main\\resources\\tmdb-movies.csv"
  var movies: mutable.MutableList[Movie] = mutable.MutableList[Movie]()

  loadMovies()
  val ends = Instant.now()
  println("Data-layer: " + Duration.between(starts, ends).toMillis() + " ms")

  private def loadMovies(): Unit = {
    //val resStream: InputStream = this.getClass.getResourceAsStream(filename)
    //val res: URL = this.getClass.getResource(filename)
    val reader: CSVReader = CSVReader.open(new File(path))
    val all: List[List[String]] = reader.all()

    for (row <- all.slice(1, all.length)) {
      movies.+=(new Movie(row(5), row(13).split('|').toList, row(17).toFloat))
    }
  }

  def getMoviesByGenre(genre: String): List[Movie] = {
    val starts = Instant.now()

    val res = movies.filter(_.genres.contains(genre)).toList

    val ends = Instant.now()
    println("Data-layer calc-score: " + Duration.between(starts, ends).toMillis + " ms")

    res
  }

  def getGenres: List[String] = {
    var genres: List[String] = movies.flatMap(_.genres).toSet.toList
    genres.filter(genre => genre!="")  // Remove empty genres if there are any
  }
}

/*object DataLayer {
  def main(args: Array[String]): Unit = {
    println("WHY?")
    new DataLayer().getGenres
  }
}*/

File: experiment1/datalayer/se/alejo720/experiment1rq1/datalayer/datalayer.py
# import sys
# import os
# PATH = os.path.dirname(os.path.abspath(__file__))
# print(PATH)
# #print(os.path.abspath(PATH))
# print(os.path.realpath(__file__))
# sys.path.insert(0, PATH)

import csv
from java.time import Duration, Instant
from se.alejo720.experiment1rq1.interfaces import DataLayerType

print("Running datalayer file from project!")

class Movie:
    def __init__(self, title, genres, score):
        self.title = title
        self.genres = genres
        self.score = score

class DataLayer(DataLayerType):
    FILENAME = "C:/Users/AlexanderJ/Google Drive/Universitet/Exjobb/intellijidea/datalayer/se/alejo720/experiment1rq1/datalayer/resources/tmdb-movies.csv"
    def __init__(self):
        starts = Instant.now()
        self.movies = []
        self.__load_movies()
        ends = Instant.now()
        print("Data-layer: {0} ms".format(Duration.between(starts, ends).toMillis()))

    def __load_movies(self):
        print "Loading movies from", self.FILENAME
        with open(self.FILENAME, "r") as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=',')
            for index, row in enumerate(csv_reader):
                if index == 0: continue # Skip first row
                self.movies.append(Movie(row[5], row[13].split('|'), float(row[17])))

    def oldgetMovieScores(self, genre, return_title = False):
        movies = []
        for movie in self.movies:
            if genre in movie.genres:
                if return_title:
                    movies.append(movie.title + "|" + str(movie.score))
                else:
                    movies.append(movie.score)
        return movies

    def getMovieScores(self, genre, return_title = False):
        starts = Instant.now()

        tmp = map(lambda m: m.score, list(filter(lambda m: genre in m.genres, self.movies)))

        ends = Instant.now()
        print("Data-layer calc-score: {0} ms".format(Duration.between(starts, ends).toMillis()))
        return tmp

    def getGenres(self):
        genres = list(set([item for sublist in map(lambda m: m.genres, self.movies) for item in sublist]))
        genres.remove("")
        print("genres", genres)
        return genres

    def test(self, t1, t2):
        return "2" + t1 + t2

    def test(self, t1):
        return "1" + t1


        """
        genres = []
        for movie in self.movies:
            for genre in movie.genres:
                if genre not in genres:
                    genres.append(genre)
        genres.remove("")
        return genres
        """
